### PR TITLE
AgentEval Offline Integration

### PR DESCRIPTION
**Pull request by jluey1**
## Related issue number

See issue #2162 
## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/agentchat/contrib/critic_agent.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+from typing import Callable, Dict, Literal, Optional, Union
+
+from autogen.agentchat.conversable_agent import ConversableAgent
+from autogen.runtime_logging import log_new_agent, logging_enabled
+
+
+class CriticAgent(ConversableAgent):
```
Comment (BeibinLi) [inline autogen/agentchat/contrib/critic_agent.py:7]: Lint question: do you think it is better to combine the critic agent, quantifier agent, and subcritic agent into one file, such as `eval_agents.py`?

File: autogen/agenteval/criterion.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import json
+
+
+class Criterion:
+    """
+    A class that represents a criterion for agent evaluation.
+    """
+
+    def __init__(self, name: str, description: str, accepted_values: [str], sub_criteria=None):
```
Comment (BeibinLi) [inline autogen/agenteval/criterion.py:9]: We use the format: `accepted_values: List[str]`

File: autogen/agenteval/criterion.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        self.name = name
+        self.description = description
+        self.accepted_values = accepted_values
+        self.sub_criteria = sub_criteria
+
+    def to_json(self):
+        """
+        Create a json object from the criterion.
+        """
+        return {self.name: {"description": self.description, "accepted_values": self.accepted_values}}
```
Comment (BeibinLi) [inline autogen/agenteval/criterion.py:25]: Should we also return `sub_criteria` here, such as a recursion fashion?

```python
return {self.name:   {"description": self.description, "accepted_values": self.accepted_values,
                                 "sub_criteria": [x.to_json() for x  in self.sub_criteria]}
        }
```

File: autogen/agenteval/criterion.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    """
+    A class that represents a criterion for agent evaluation.
+    """
+
+    def __init__(self, name: str, description: str, accepted_values: [str], sub_criteria=None):
+        """
+        args:
+        - name (str): The name of the criterion.
+        - description (str): The description of the criterion.
+        - accepted_values ([str]): The list of accepted values for the criterion.
```
Comment (BeibinLi) [inline autogen/agenteval/criterion.py:14]: `sub_criteria` is not described in docstring, and the accepted type is undefined.

File: notebook/agenteval_cq_math.ipynb [hunk 5]
```diff
@@ -168,52 +112,39 @@
...
+    "        }\n",
+    "    )\n",
+    ")\n",
     "\n",
-    "sys_msg = f\"\"\"Task: {task[\"name\"]}.\n",
-    "Task description: {task[\"description\"]}\n",
-    "Task successful example: {task[\"successful_response\"]}\n",
-    "Task failed example: {task[\"failed_response\"]}\n",
-    "\"\"\""
+    "criteria = generate_criteria(task=task, llm_config={\"config_list\": config_list})"
```
Comment (BeibinLi) [inline notebook/agenteval_cq_math.ipynb:147]: I feel we should visualize the criteria for readers to understand the setup better.

Maybe `print(criteria.to_json())` here would help.

File: notebook/agenteval_cq_math.ipynb
Comment (BeibinLi) [inline notebook/agenteval_cq_math.ipynb:1]: The last few cells are not ran, and the visualization results are missing.

File: autogen/agentchat/contrib/critic_agent.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    DEFAULT_DESCRIPTION = "An AI agent for creating list criteria for evaluating the utility of a given task."
+
+    def __init__(
+        self,
+        name="critic",
+        system_message: Optional[str] = DEFAULT_SYSTEM_MESSAGE,
+        llm_config: Optional[Union[Dict, bool]] = None,
+        max_consecutive_auto_reply: Optional[int] = None,
+        human_input_mode: Optional[str] = "NEVER",
+        description: Optional[str] = DEFAULT_DESCRIPTION,
```
Comment (sonichi) [inline autogen/agentchat/contrib/critic_agent.py:27]: For the arguments that you don't need to alter the default, don't list them explicitly. That will make it easier to maintain.
Also, the current way of writing the constructor will remove a few kwargs and stop them from being passed to super(). Is that intentional?

**Comment by jluey1** [comment]
@qingyun-wu , do you mind taking a look at this PR while Chi is out?

**Comment by sonichi** [comment]
Could you add a test to cover the newly added code? Depending on whether the test requires openai, you can add it to contrib-tests.yml or contrib-openai.yml. It takes some learning to understand how to write the tests. Please check other contrib tests and how they are added to these workflows.

File: test/agentchat/contrib/agent_eval/test_agent_eval.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                "description": "Given any question, the system needs to solve the problem as consisely and accurately as possible",
+                "successful_response": response_successful,
+                "failed_response": response_failed,
+            }
+        )
+    )
+
+
+@pytest.mark.skipif(
+    skip_openai,
```
Comment (sonichi) [inline test/agentchat/contrib/agent_eval/test_agent_eval.py:69]: To run this test in CI, we need to (1) create a new PR from a branch in the upstream repo; (2) add a section in contrib-openai.yml to run the test.

File: test/agentchat/contrib/agent_eval/test_criterion.py
Comment (sonichi) [inline test/agentchat/contrib/agent_eval/test_criterion.py:1]: To run these tests in CI, we need to add a section in contrib-tests.yml.

File: test/agentchat/contrib/agent_eval/test_agent_eval.py (hunk 7 above)
Comment (jluey1) [reply test/agentchat/contrib/agent_eval/test_agent_eval.py:69]: @sonichi can you clarify what you mean by (1)? is this a new PR entirely or just merging in changes from a different branch?

Comment (sonichi) [reply test/agentchat/contrib/agent_eval/test_agent_eval.py:69]: A new PR with the same changes as in this PR.

**Review by julianakiseleva — APPROVED** [review]
Looks good to me

**Comment by jluey1** [comment]
Closing in order to make a new PR from agenteval_integration into main

**Pull request closed without merging**